---
title: "Allelic expression analysis with Salmon and Swish"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true
bibliography: library.bib
vignette: |
  %\VignetteIndexEntry{2. Allelic expression analysis with Salmon and Swish}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- run this document with rmarkdown::render("swish.Rmd") -->

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=FALSE, dev="png",
                      message=FALSE, error=FALSE, warning=FALSE)
```

# Allelic analysis

# Linking transcripts to TSS

```{r}
suppressPackageStartupMessages(library(ensembldb))
library(EnsDb.Hsapiens.v86)
library(fishpond)
edb <- EnsDb.Hsapiens.v86
t2g <- makeTx2Tss(edb)
mcols(t2g)[,c("tx_id","group_id")]
```

# Importing allelic counts

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
```

# Testing for allelic imbalance across samples

```{r}
set.seed(1)
y <- makeSimSwishData(allelic=TRUE)
```

A hidden code chunk is used to add ranges from the *EnsDb* to the
simulated data set. For a real dataset, the ranges would be added
either by `importAllelicCounts` (if using `tx2gene`) or could be added
manually for transcript- or gene-level analysis, using the
`rowRanges<-` setter function. The ranges are only needed for the
`plotAllelicGene` plotting function below.

```{r echo=FALSE}
# hidden chunk to add ranges to the `se`
tss <- t2g[!duplicated(t2g$group_id)]
names(tss) <- tss$group_id
mcols(tss) <- mcols(tss)[,c("tx_id","gene_id","tss","group_id")]
# slow...
#tx_id <- CharacterList(split(t2g$tx_id,t2g$group_id))
#tss$tx_id <- tx_id[names(tss)]
tab <- table(tss$gene_id)
tss$ntss <- as.numeric(tab[tss$gene_id])
tss <- tss[tss$ntss > 1 & tss$ntss < 5 & seqnames(tss) == "1"]
tss <- tss[order(tss$gene_id),]
tss <- tss[43:1042]
# swap 2nd and 3rd isoform of first gene
tss[2:3] <- tss[3:2] 
rowRanges(y) <- tss
```

We can already plot a heatmap of allelic ratios, before performing
statistical testing.

```{r fig.dim=c(7,5)}
gene <- rowRanges(y)$gene_id[1]
idx <- mcols(y)$gene_id == gene
plotAllelicHeatmap(y, idx=idx)
```

```{r}
y <- labelKeep(y)
y <- swish(y, x="allele", pair="sample")
```

# Plotting results

```{r fig.dim=c(7,7)}
par(mfrow=c(2,1))
plotInfReps(y, idx=1, x="allele", cov="sample")
plotInfReps(y, idx=2, x="allele", cov="sample")
```

```{r fig.dim=c(7,7)}
plotAllelicGene(y, gene, edb)
```

```{r fig.dim=c(7,7)}
plotAllelicGene(y, gene, edb,
                transcriptAnnotation="transcript")
```

# Testing for dynamic allelic imbalance

```{r}
set.seed(1)
y <- makeSimSwishData(dynamic=TRUE)
```

```{r fig.dim=c(7,5)}
plotAllelicHeatmap(y, idx=1:3)
```

```{r}
y <- labelKeep(y)
y <- swish(y, x="allele", pair="sample", cov="time", cor="pearson")
```

```{r}
mcols(y)[1:2,c("stat","qvalue")]
```

```{r fig.dim=c(7,7)}
y <- computeInfRV(y)
par(mfrow=c(2,1))
plotInfReps(y, idx=1, x="time", cov="allele", shiftX=.01)
plotInfReps(y, idx=2, x="time", cov="allele", shiftX=.01)
```

```{r fig.dim=c(7,7)}
plotInfReps(y, idx=1, x="time", cov="allele", shiftX=.01)
dat <- data.frame(
  time = y$time[1:10],
  a2 = assay(y, "mean")[1,y$allele=="a2"],
  a1 = assay(y, "mean")[1,y$allele=="a1"])
lines(lowess(dat[,c(1,2)]), col="dodgerblue")
lines(lowess(dat[,c(1,3)]), col="goldenrod4")
```
