---
title: "Fishpond: differential gene expression and differential transcript expression with inferential replicates"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
author: "Anqi Zhu, Rob Patro, Michael Love"
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_float: true
vignette: |
  %\VignetteIndexEntry{Fishpond}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- run this document with library(rmarkdown); render("fishpond.Rmd") -->

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=FALSE, error=FALSE, warning=FALSE)
```

```{r}
library(tximeta)
suppressPackageStartupMessages(library(SummarizedExperiment))
dir <- "../../quants"
subdir <- rep(c("bias","unif"), each=12)
nms <- rep(as.vector(t(outer(1:6,1:2,function(x,y) paste0(x,"_",y)))),2)
files <- file.path(dir, subdir, nms, "quant.sf")
all(file.exists(files))
condition <- factor(sub(".*_(.)","\\1",nms))
batch <- factor(rep(1:2,each=12))
names <- paste0(nms, "_b", batch)
coldata <- data.frame(files, names, condition, batch)
se <- tximeta(coldata)
#y <- se
#save(y, file="y_full.rda")
```

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
y <- summarizeToGene(se)
y2 <- summarizeToGene(se, varReduce=TRUE)
# scaling from counts to TPM and back to lib corrected counts
y <- scaleInfReps(y)
# labelling sufficiently expressed transcripts/genes
y <- labelKeep(y)
y <- y[,y$batch == 1]
y2 <- y2[,y2$batch == 1]
infCV <- sqrt(assays(y2)[["variance"]])/(assays(y2)[["counts"]]+0.5)
meanInfCV <- rowMeans(infCV)
mcols(y)$InfCV <- meanInfCV
y <- y[mcols(y)$keep,] 
#save(y, file="y.rda") # 160 Mb
#load("y.rda")
```

```{r}
y <- swish(y, x="condition")
```

```{r}
y <- swish(y, x="condition", cov="batch")
```

```{r}
load("../../quants/simulate.rda")
sig <- rownames(y)[mcols(y)$locfdr < .05]
dte.txps <- names(iso.dtu)[iso.dtu | iso.dte | iso.dge]
prop.table(table(sig %in% dte.txps))
#
sig <- rownames(y)[mcols(y)$qvalue < .1]
dte.txps <- names(iso.dtu)[iso.dtu | iso.dte | iso.dge]
prop.table(table(sig %in% dte.txps))
mcols(y)$sig <- mcols(y)$qvalue < .1
mcols(y)$fp <- NA
mcols(y)$fp[mcols(y)$sig] <- !sig %in% dte.txps
mcols(y)$cutICV <- cut(mcols(y)$InfCV, c(-1,.05,.25,.5,1,Inf))
table(mcols(y)$cutICV, mcols(y)$fp)
round(prop.table(table(mcols(y)$cutICV, mcols(y)$fp),1),3)
```

```{r}
sig <- rownames(y)[mcols(y)$qvalue < .1]
dge <- union(dge.genes, dte.genes)
prop.table(table(sig %in% dge))
mcols(y)$sig <- mcols(y)$qvalue < .1
mcols(y)$fp <- NA
mcols(y)$fp[mcols(y)$sig] <- !sig %in% dge
mcols(y)$cutICV <- cut(mcols(y)$InfCV, c(-1,.05,.25,.5,1,Inf))
table(mcols(y)$cutICV, mcols(y)$fp)
round(prop.table(table(mcols(y)$cutICV, mcols(y)$fp),1),3)
```

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
load("y_full.rda")
x <- model.matrix(~batch + condition, colData(y))
coef <- "condition2"
mcols(y)$keep <- rowSums(assays(y)[["counts"]] >= 10) >= 6
y2 <- y[1:2000,]
y2 <- deswish(y2, x, coef)
```

